libclc/native_cpu-unknown-linux/libspirv/integer/popcount.cl

```diff
@@ -2,12 +2,25 @@
 #include <clcmacro.h>
 #include <spirv/spirv.h>
 
+// We can't use __builtin_popcountg because it supports only unsigned
+// types, and we can't use __builtin_popcount because the implicit cast
+// to int doesn't work due to sign extension, so we explictly cast to
+// the unsigned type.
+#define DEF_POPCOUNT_HELPER(TYPE, UTYPE) \
+_CLC_OVERLOAD TYPE __popcount_helper(TYPE c) { \
+  return __builtin_popcountg((UTYPE)c); \
```

Comment:
I might be missing something, but this patch sees incorrect.

According to OpenCL's spec ([ver 1.2 Last Revision Date: 11/14/12 Page 258 6.12.3 Integer Functions](https://registry.khronos.org/OpenCL/specs/opencl-1.2.pdf)) `popcount` is defined for all integer types (including the signed versions and vectors), with this patch we would never be able to distinguish between:
```
popcount(-5)
popcount(5u)
```

Reply:
We discussed it offline and indeed, I was missing something :)

Reply:
We've also discussed a type punning through a pointer dereference here, instead of a simple cast.